Experimental Prometheus Remote Write guide (#943)
grafana/k6-docs: src/data/markdown/translated-guides/en/03 Results output/200 Real-time/00 Prometheus.md

---

<ExperimentalBlockquote />

Prometheus remote write is a protocol with a defined [specification](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit).
It has multiple implementations.
For example, one option is to store the metrics in [Prometheus](https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver).
You can find other implementations in the Prometheus' [Integrations](https://prometheus.io/docs/operating/integrations) guide.

With the Prometheus remote write output, k6 can send test-result metrics to a Prometheus remote write endpoint.
The output during the `k6 run` execution gets all the generated data points for the [built-in  k6 metrics](/using-k6/metrics/).
It then generates the equivalent Prometheus remote write time series.

## Metrics mapping

All the k6 metric types are converted into an equivalent Prometheus remote write type:

| k6 | Prometheus | Name label |
|----|------------| ---------- |
| Counter | Counter | `k6_*_total`
| Gauge | Gauge | `k6_*_<unit-suffix>` |
| Rate | Gauge | `k6_*_rate` |
| Trend | Gauges / Native Histogram | `k6_*_<unit-suffix>` |

### Trend

> The obvious conversion with a classic Prometheus Histogram is not convenient because k6 can't determine the fixed buckets in advance.

Author: >The first solution is an "hack" and technically is Trend-to-Counters-and-Gauges

Also, we aren't doing Trend to "classic" Histogram because it would require the user to specify the static buckets and it would create a bad UX.

> The obvious conversion with a classic Prometheus Histogram is not convenient because k6 can't determine the fixed buckets in advance.

Reviewer: ```
## Trend metric conversion

By default, k6 trend metrics convert to primitive counters and gauges. To convert trend metrics to high-fidelity histograms, you can use Prometheus Native histogram instead.

Because k6 can't easily determine fixed buckets in advance, k6 metrics can't export to the original Prometheus histograms.
So, the output maps a trend metric into primitive counter and gauges.
Each value represents a math function (count, sum, min, max, avg, med, p(x)).
This mapping has following drawbacks:
- It is impossible to aggregate some gauge values (especially percentiles).
- It uses a memory-expensive k6 data structure.

To resolve these limitations, you can map a Trend as a [Prometheus Native Histogram].(https://prometheus.io/docs/concepts/metric_types/#histogram).
To do this, you can use the `K6_PROMETHEUS_RW_TREND_AS_NATIVE_HISTOGRAM=true` environment variable (or one of the other ways).
The output then converts all the trend types into a dedicated Native Histogram.

Mapping trends as native histograms gives better you more efficient storage and preciser queries.
The drawback is that **the feature is experimental, released in Prometheus v2.40.0**.

<Blockquote mod="note" title="">

Native Histogram is an experimental feature, so it has to be enabled ([--enable-feature=native-histograms](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms)). Note that other remote write implementations don't support it yet.

</Blockquote>

```

What do you think about that? Writing that made it clearer to me, but I don't know about other readers.

---
title: 'Prometheus remote write'
slug: '/results-output/real-time/prometheus'
excerpt: 'Use the Prometheus remote write integration to send test results to any Prometheus remote write endpoint.'
---

<ExperimentalBlockquote />

> Prometheus remote write is a protocol with a defined [specification](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit).
> It has multiple implementations.
> For example, you can store the metrics in [Prometheus](https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver).
> For other implementations, check the Prometheus [Integrations](https://prometheus.io/docs/operating/integrations) guide.

Reviewer: ```suggestion
Prometheus remote write is a protocol to make it possible to reliably propagate data in real-time from a sender to a receiver.
It has a defined [specification](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit)
and multiple implementations.
For example, you can store the metrics in [Prometheus](https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver).
For other implementations, check the Prometheus [Integrations](https://prometheus.io/docs/operating/integrations) guide.
```

It then generates the equivalent Prometheus remote write time series.

## About metrics mapping

All k6 metric types are converted into an equivalent Prometheus remote write type:

| k6      | Prometheus                | Name label           |
|---------|---------------------------|----------------------|
| Counter | Counter                   | `k6_*_total`         |
| Gauge   | Gauge                     | `k6_*_<unit-suffix>` |
| Rate    | Gauge                     | `k6_*_rate`          |
| Trend   | Gauges / Native Histogram | `k6_*_<unit-suffix>` |

### Trend metric conversion

By default, k6 trend metrics convert to primitive counters and gauges.
To convert trend metrics to high-fidelity histograms, you can use Prometheus Native histogram instead.

Because k6 can't easily determine fixed buckets in advance, k6 metrics can't export to the original Prometheus histograms.
So, the output maps a trend metric into primitive counter and gauges.
Each value represents a math function (count, sum, min, max, avg, med, p(x)).
This mapping has following drawbacks:
- It is impossible to aggregate some gauge values (especially percentiles).
- It uses a memory-expensive k6 data structure.

> To resolve these limitations, you can map a trend as a [Prometheus Native Histogram].(https://prometheus.io/docs/concepts/metric_types/#histogram).

Author: ```suggestion
To resolve these limitations, you can map a trend as a [Prometheus Native Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram).
```

---

> title: 'Prometheus Remote Write'

Reviewer: Oh I see. Even there own docs don't capitalize it though. 

I think with some tactical `find` and `sed` operations, we can standardize it pretty easily either way.

excerpt: 'Prometheus Remote Write integration allows to send test results to any Prometheus Remote Write endpoint.'
---

<Blockquote mod="attention" title="Experimental">

> Prometheus Remote Write Output is an experimental feature and breaking change to the API or the generated time series could be introduced in the future. 

Author: This is nice

> Prometheus Remote Write Output is an experimental feature and breaking change to the API or the generated time series could be introduced in the future. 

Author: ```suggestion
<ExperimentalBlockquote />
```

</Blockquote>

k6 supports sending test result metrics to a Prometheus Remote Write endpoint via the Prometheus Remote Write Output.

> Prometheus Remote Write is a protocol with a defined [specification](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit) and it has multiple implementations, one option with this support is storing the metrics in [Prometheus](https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver); others can be found in the Prometheus' [Integrations](https://prometheus.io/docs/operating/integrations) guide.

Author: Unfortunately, this is the unique complete source that I've found. All the rest just focus on specific parts.

The Output during the `k6 run` execution gets all the generated data points for the defined k6's [builtin metrics](/using-k6/metrics/#built-in-metrics) then it generates the equivalent Prometheus's Remote Write time series.

## How to use

A up and running Remote Write implementation is required exposing an endpoint reachable by k6's process. Check the guide below how to run Prometheus as a Remote write agent if you need one.
Then a test run with the Output can be used as follows:

```bash
k6 run -o experimental-prometheus-rw script.js
```
 
All the time series generated and sent by the Output get the prefix `k6_` attached to their names. In case of Prometheus, they can be seen in its UI:

> ![k6 metrics as seen in the Prometheus UI](images/Prometheus/prom.png)

Author: Autospot, this image must be updated

## Metrics mapping

All the k6 metric types are converted into an equivalent Prometheus Remote Write type:

| k6 | Prometheus | Name label |
|----|------------| ---------- |
| Counter | Counter | `k6_*_total`
| Gauge | Gauge | `k6_*_<unit-suffix>` |
| Rate | Gauge | `k6_*_rate` |
| Trend | Gauges / Native Histogram | `k6_*_<unit-suffix>` |

### Trend

The obvious conversion with a classic Prometheus Histogram is not convenient because k6 can't determine the fixed buckets in advance, so the Output maps a Trend metric by default into primitive Counter and Gauges where each value represents a math function (count, sum, min, max, avg, med, p(x)).
Mapping Trend by Stats has the following cons:
* It is impossible to aggregate some Gauge's value (especially the percentiles).

> * It uses a memory-expensive k6's data structure.

Reviewer: ```suggestion
* It uses a memory-expensive k6 data structure.
```

> The previous points can be resolved by mapping Trend as [Prometheus Native Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram). Enabling the conversion by the `K6_PROMETHEUS_RW_TREND_AS_NATIVE_HISTOGRAM=true` environment variable (or one of the other ways), then the Output converts all the Trend types into a dedicated Native Histogram.

Reviewer: ```suggestion
To resolve these limitations, you can map a Trend as a [Prometheus Native Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram).
You can enable conversion with the `K6_PROMETHEUS_RW_TREND_AS_NATIVE_HISTOGRAM=true` environment variable (or one of the other ways).
The output then converts all the trend types into a dedicated Native Histogram.
```

> Native Histogram is a Prometheus' experimental feature, so it has to be enabled (`--enable-feature=native-histograms`). Note that other Remote Write implementations don't support it yet.

Reviewer: ```suggestion
Native Histogram is an experimental feature, so it has to be enabled (`--enable-feature=native-histograms`). Note that other Remote Write implementations don't support it yet.
```

### Naming convention

The Output maps the metrics into time series with Name labels respecting as much as possible the naming [best practices](https://prometheus.io/docs/practices/naming) defined by the Prometheus' project:

* All the time series are prefixed with the `k6_` namespace.

> * All the time series contains when known a suffix with the Base unit used for Sample's values.

Reviewer: > * All the time series contains when known a suffix with the Base unit used for Sample's values.

A little confused here. Perhaps, 

> All time series are suffixed with the base unit of the sample value (if k6 knows what the base unit is).

?

Author: ```suggestion
* All time series are suffixed with the base unit of the sample value (if k6 knows what the base unit is).
```

> * Trend's stats and Rate contains the relative suffix for making them better discoverable.

Reviewer: ```suggestion
* Trends and Rates have the relative suffixes, to make them more discoverable.
```
